pandas/_libs/tslibs/resolution.pyx

```diff
@@ -0,0 +1,649 @@
+
+    @classmethod
+    def get_reso(cls, resostr):
+        """
+        Return resolution str against resolution code.
+
+        Example
+        -------
+        >>> Resolution.get_reso('second')
+        2
+
+        >>> Resolution.get_reso('second') == Resolution.RESO_SEC
+        True
+        """
+        return cls._str_reso_map.get(resostr, cls.RESO_DAY)
+
+    @classmethod
+    def get_freq_group(cls, resostr):
+        """
+        Return frequency str against resolution str.
+
+        Example
+        -------
+        >>> f.Resolution.get_freq_group('day')
+        4000
+        """
+        return get_freq_group(cls.get_freq(resostr))
+
+    @classmethod
+    def get_freq(cls, resostr):
+        """
+        Return frequency str against resolution str.
+
+        Example
+        -------
+        >>> f.Resolution.get_freq('day')
+        'D'
+        """
+        return cls._reso_freq_map[resostr]
+
+    @classmethod
+    def get_str_from_freq(cls, freq):
+        """
+        Return resolution str against frequency str.
+
+        Example
+        -------
+        >>> Resolution.get_str_from_freq('H')
+        'hour'
+        """
+        return cls._freq_reso_map.get(freq, 'day')
+
+    @classmethod
+    def get_reso_from_freq(cls, freq):
+        """
+        Return resolution code against frequency str.
+
+        Example
+        -------
+        >>> Resolution.get_reso_from_freq('H')
+        4
+
+        >>> Resolution.get_reso_from_freq('H') == Resolution.RESO_HR
+        True
+        """
+        return cls.get_reso(cls.get_str_from_freq(freq))
+
+    @classmethod
+    def get_stride_from_decimal(cls, value, freq):
+        """
+        Convert freq with decimal stride into a higher freq with integer stride
+
+        Parameters
+        ----------
+        value : integer or float
+        freq : string
+            Frequency string
+
+        Raises
+        ------
+        ValueError
+            If the float cannot be converted to an integer at any resolution.
+
+        Example
+        -------
+        >>> Resolution.get_stride_from_decimal(1.5, 'T')
+        (90, 'S')
+
+        >>> Resolution.get_stride_from_decimal(1.04, 'H')
+        (3744, 'S')
+
+        >>> Resolution.get_stride_from_decimal(1, 'D')
+        (1, 'D')
+        """
+        if np.isclose(value % 1, 0):
+            return int(value), freq
+        else:
+            start_reso = cls.get_reso_from_freq(freq)
+            if start_reso == 0:
+                raise ValueError("Could not convert to integer offset "
+                                 "at any resolution")
+
+            next_value = cls._reso_mult_map[start_reso] * value
+            next_name = cls._reso_str_bump_map[freq]
+            return cls.get_stride_from_decimal(next_value, next_name)
+
+
+# ----------------------------------------------------------------------
+# Frequency Inference
+
+
+def unique_deltas(ndarray[int64_t] arr):
```

Comment:
is this actually used elsewhere? if so should fix that, this should be a cdef method of _FrequencyInferer  (or at the very lease cdef only)

Reply:
Nope, just in resolution.  Will cdef for now.

Reply:
Related: there's an import of `pandas.core.algorithms.unique`.  Do you know if there is a cython version of that we can use instead?  The uses are all for fairly small integer indexes:

```
if len(unique(self.fields['M'])) > 1:
weekdays = unique(self.index.weekday)
week_of_months = unique((self.index.day - 1) // 7)
```

In fact, if these are all being called on `Index` objects, then the `Int64Index.unique` method might make sense.